# Pectin effects on gut microbiota (QIIME2 & R)

This repository provides the scripts and instructions necessary to reproduce the analysis of 16S rRNA gene sequencing  data used to investigate the effects of pectin-based prebiotic fiber derived from olive oil by-products on gut microbiota of Alzheimer's disease and obese mouse models. The analysis includes data preprocessing, quality control, and statistical analysis to explore the microbial community structure and diversity following the pectin treatemnt in the diverse groups.  To follow this repository you will need to have installed QIIME2 and R.

## QIIME 2 Installation

To install QIIME 2 in a Conda environment, follow these steps:

1. **Download and Install Miniconda**:
   If not already installed, download and install Miniconda from the [official site](https://docs.conda.io/en/latest/miniconda.html).

2. **Set up the QIIME 2 Conda Environment**:
   Execute the following commands:

   ```sh
   # Download the QIIME 2 environment file
   wget https://data.qiime2.org/distro/core/qiime2-2023.2-py38-linux-conda.yml
   # Create a new conda environment named 'qiime2'
   conda env create -n qiime2 --file qiime2-2023.2-py38-linux-conda.yml
   
## Scripts on this repository

* The **preprocessing.sh** file is a Bash script containing all the code necessary to perform data importing to QIIME2 and the denoising process.
  * The script takes as an input 16S sequencing data on **Casava 1.8 demultiplexed paired-end format**. If you want to follow this script make sure your data is in this format
    or else modify the **qiime tools import** command from the script accordingly to match your data type. You can look for more information on this topic on the [QIIME2 official tutorial](https://docs.qiime2.org/2024.5/tutorials/importing/).
  * The script was ran on a linux remote server from CICA (_Centro Informatico Cientifico de Andalucia_) with a SSH protocol with the following features:
    * **Operating system:** Ubuntu 20.04.2 LTS
    * **Kernel version:** 5.4.0-146-generic
    * **CPU information:** AMD Opteron(tm) Processor 6378 with 32 CPUs
    * **RAM memory:** 62GB
* The **pectin_effects_on_gut_microbiota.R** file is an R script containing the code to import the artifacts generated by QIIME2 and perform further downstream analysis.
   * Every package needed for the R downstream analysis is specified and installed at the beginning of the R script.

# Contact

This study is part of a project carried out by the **Immunonutrition and Immunometabolism** group (CTS1074) from the University of Seville

For any questions or issues, please open an issue on GitHub or contact **anavarrorubira@gmail.com**
